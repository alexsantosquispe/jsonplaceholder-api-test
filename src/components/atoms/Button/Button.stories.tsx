import type { Meta, StoryObj } from '@storybook/react';

import { Button, ButtonProps } from './Button';
import { Wrapper } from '../../../utils/stories-utils';

const meta = {
  title: 'Atoms/Button',
  component: Button,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof Button>;

export const Default: Story = (args: Partial<ButtonProps>) => {
  return (
    <div className="grid grid-cols-4 gap-4">
      <Wrapper className="flex-col">
        <Button label={args.label} />
        <span className="text-sm">Default</span>
      </Wrapper>
      <Wrapper className="flex-col">
        <Button label={args.label} isDisabled />
        <span className="text-sm">Disabled</span>
      </Wrapper>
      <Wrapper className="flex-col">
        <Button label="Secondary" isSecondary />
        <span className="text-sm">Secondary</span>
      </Wrapper>
      <Wrapper className="flex-col">
        <Button label="Secondary" isSecondary isDisabled />
        <span className="text-sm">Disabled</span>
      </Wrapper>
    </div>
  );
};

Default.args = {
  label: 'Button',
  onClick: () => alert('Button clicked!')
};
