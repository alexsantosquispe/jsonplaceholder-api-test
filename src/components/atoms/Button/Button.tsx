import cn from 'clsx';
import { ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';

export interface ButtonProps {
  label?: string;
  type?: 'button' | 'submit';
  icon?: ReactNode;
  onClick?: () => void;
  isSecondary?: boolean;
  isDisabled?: boolean;
  className?: string;
  ariaLabel?: string;
}

export const Button = ({
  label,
  type = 'button',
  onClick,
  icon = null,
  isDisabled = false,
  isSecondary = false,
  className = '',
  ariaLabel
}: ButtonProps) => {
  return (
    <button
      role="button"
      aria-disabled={isDisabled}
      aria-label={ariaLabel}
      className={twMerge(
        'flex cursor-pointer items-center justify-center gap-1 rounded-lg px-4 py-2 text-sm font-medium transition-colors duration-150 ease-in-out',
        cn({
          'bg-primary dark:text-primary text-white shadow-xs dark:bg-white/90':
            !isSecondary,
          'hover:bg-primary/80 dark:hover:bg-white/80': !isDisabled,
          'text-primary/80 hover:text-primary border border-black/10 bg-transparent shadow-xs hover:bg-gray-100 dark:border-white/15 dark:text-white dark:hover:bg-white/15':
            isSecondary,
          'text-primary/40 hover:text-primary/40 cursor-not-allowed border-black/5 hover:bg-transparent dark:border-white/10 dark:text-white/25 dark:hover:bg-transparent':
            isSecondary && isDisabled,
          'bg-primary/65 cursor-not-allowed dark:bg-white/25':
            !isSecondary && isDisabled
        }),
        className
      )}
      disabled={isDisabled}
      type={type}
      onClick={onClick}
    >
      {icon && <span className="size-4">{icon}</span>}
      {label && label}
    </button>
  );
};
