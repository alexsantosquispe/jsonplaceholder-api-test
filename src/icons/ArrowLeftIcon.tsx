import { twMerge } from 'tailwind-merge';

import { SvgIconProps } from '../types';

export const ArrowLeftIcon = ({ className = '' }: SvgIconProps) => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      fill="none"
      viewBox="0 0 24 24"
      strokeWidth={1.5}
      stroke="currentColor"
      className={twMerge('size-6', className)}
    >
      <path
        strokeLinecap="round"
        strokeLinejoin="round"
        d="M10.5 19.5 3 12m0 0 7.5-7.5M3 12h18"
      />
    </svg>
  );
};
