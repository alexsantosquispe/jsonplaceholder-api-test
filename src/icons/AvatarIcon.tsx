import { twMerge } from 'tailwind-merge';
import { BaseIconProps } from '../types';

export const AvatarIcon = ({ className = '' }: BaseIconProps) => {
  return (
    <div className="relative h-10 w-10 overflow-hidden rounded-full bg-gray-100">
      <svg
        className={twMerge(
          'absolute -left-1 h-12 w-12 text-gray-400',
          className
        )}
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          fillRule="evenodd"
          d="M10 9a3 3 0 100-6 3 3 0 000 6zm-7 9a7 7 0 1114 0H3z"
          clipRule="evenodd"
        ></path>
      </svg>
    </div>
  );
};
