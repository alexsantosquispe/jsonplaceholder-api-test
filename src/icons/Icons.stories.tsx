import { Meta } from '@storybook/react';
import { useState } from 'react';

import * as Icons from './index';
import { INPUT_FIELD_STYLES } from '../constants/styles';

export default {
  title: 'Icons'
} as Meta;

export const IconGallery = () => {
  const [filter, setFilter] = useState('');

  const filteredIcons = Object.entries(Icons).filter(([name]) =>
    name.toLowerCase().includes(filter.toLowerCase())
  );

  return (
    <div className="space-y-6 p-4">
      <div className="flex w-1/2 flex-col gap-y-1">
        <label>Icons Gallery</label>
        <input
          placeholder="Search icons..."
          className={INPUT_FIELD_STYLES}
          value={filter}
          onChange={(e) => setFilter(e.target.value)}
        />
      </div>

      <div className="flex flex-wrap gap-4">
        {filteredIcons.map(([name, IconComponent]) => (
          <div
            key={name}
            className="flex h-[10rem] w-[10rem] flex-col items-center justify-center gap-4 rounded-lg border border-gray-200"
          >
            <IconComponent className="size-6" />
            <span className="text-xs font-medium text-gray-600">{name}</span>
          </div>
        ))}
      </div>
    </div>
  );
};
