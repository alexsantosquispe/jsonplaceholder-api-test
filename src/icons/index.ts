import { ArrowLeftIcon } from './ArrowLeftIcon';
import { AvatarIcon } from './AvatarIcon';
import { ChatIcon } from './ChatIcon';
import { ChevronRightIcon } from './ChevronRightIcon';
import { ChevronUpDownIcon } from './ChevronUpDownIcon';
import { CloseIcon } from './CloseIcon';
import { FolderIcon } from './FolderIcon';
import { HeartIcon } from './HeartIcon';
import { MenuIcon } from './MenuIcon';
import { MoonIcon } from './MoonIcon';
import { PlusCircle } from './PlusCircle';
import { SunIcon } from './SunIcon';

export {
  ArrowLeftIcon,
  AvatarIcon,
  ChatIcon,
  ChevronRightIcon,
  CloseIcon,
  FolderIcon,
  HeartIcon,
  MenuIcon,
  MoonIcon,
  PlusCircle,
  SunIcon,
  ChevronUpDownIcon
};
