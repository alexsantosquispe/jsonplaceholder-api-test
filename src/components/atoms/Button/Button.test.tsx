import { fireEvent, render, screen } from '@testing-library/react';

import { Button } from './Button';
import { PlusCircleIcon } from '../../../icons';

describe('Button', () => {
  describe('Styles', () => {
    const onClickMock = jest.fn();
    const props = {
      icon: <PlusCircleIcon />,
      onClick: onClickMock
    };

    it('should render the component correctly for primary type', () => {
      const button = render(<Button {...props} label="Primary button" />);

      expect(button).toMatchSnapshot();
    });

    it('should render the component correctly for secondary type', () => {
      const button = render(
        <Button {...props} label="Secondary button" isSecondary />
      );

      expect(button).toMatchSnapshot();
    });

    it('should click the button', () => {
      render(<Button {...props} label="Secondary button" />);

      const button = screen.getByRole('button');

      fireEvent.click(button);

      expect(onClickMock).toHaveBeenCalledTimes(1);
    });
  });
});
