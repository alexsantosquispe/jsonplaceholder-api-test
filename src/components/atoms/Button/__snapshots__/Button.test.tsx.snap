// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Styles should render the component correctly for primary type 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-disabled="false"
        class="flex cursor-pointer items-center justify-center gap-1 rounded-lg px-4 py-2 text-sm font-medium transition-colors duration-150 ease-in-out bg-primary dark:text-primary text-white shadow-xs dark:bg-white/90 hover:bg-primary/80 dark:hover:bg-white/80"
        role="button"
        type="button"
      >
        <span
          class="size-4"
        >
          <svg
            class="size-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 9v6m3-3H9m12 0a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </span>
        Primary button
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-disabled="false"
      class="flex cursor-pointer items-center justify-center gap-1 rounded-lg px-4 py-2 text-sm font-medium transition-colors duration-150 ease-in-out bg-primary dark:text-primary text-white shadow-xs dark:bg-white/90 hover:bg-primary/80 dark:hover:bg-white/80"
      role="button"
      type="button"
    >
      <span
        class="size-4"
      >
        <svg
          class="size-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 9v6m3-3H9m12 0a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
      Primary button
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button Styles should render the component correctly for secondary type 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-disabled="false"
        class="flex cursor-pointer items-center justify-center gap-1 rounded-lg px-4 py-2 text-sm font-medium transition-colors duration-150 ease-in-out text-primary/80 hover:text-primary border border-black/10 bg-transparent shadow-xs hover:bg-gray-100 dark:border-white/15 dark:text-white dark:hover:bg-white/15"
        role="button"
        type="button"
      >
        <span
          class="size-4"
        >
          <svg
            class="size-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 9v6m3-3H9m12 0a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </span>
        Secondary button
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-disabled="false"
      class="flex cursor-pointer items-center justify-center gap-1 rounded-lg px-4 py-2 text-sm font-medium transition-colors duration-150 ease-in-out text-primary/80 hover:text-primary border border-black/10 bg-transparent shadow-xs hover:bg-gray-100 dark:border-white/15 dark:text-white dark:hover:bg-white/15"
      role="button"
      type="button"
    >
      <span
        class="size-4"
      >
        <svg
          class="size-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 9v6m3-3H9m12 0a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
      Secondary button
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
